tools/pubsys/src/repo/validate_repo/mod.rs

```diff
@@ -38,52 +39,39 @@ pub(crate) struct ValidateRepoArgs {
     validate_targets: bool,
 }
 
-/// If we are on a machine with a large number of cores, then we limit the number of simultaneous
-/// downloads to this arbitrarily chosen maximum.
-const MAX_DOWNLOAD_THREADS: usize = 16;
+/// Retrieves listed targets and attempts to download them for validation purposes.
+async fn retrieve_targets(repo: Arc<Mutex<Repository>>) -> Result<(), Error> {
+    let targets = repo.lock().await.targets().signed.targets.clone();
+    let mut tasks = JoinSet::new();
 
-/// Retrieves listed targets and attempts to download them for validation purposes. We use a Rayon
-/// thread pool instead of tokio for async execution because `reqwest::blocking` creates a tokio
-/// runtime (and multiple tokio runtimes are not supported).
-fn retrieve_targets(repo: &Repository) -> Result<(), Error> {
-    let targets = &repo.targets().signed.targets;
-    let thread_pool = rayon::ThreadPoolBuilder::new()
-        .num_threads(min(num_cpus::get(), MAX_DOWNLOAD_THREADS))
-        .build()
-        .context(error::ThreadPoolSnafu)?;
-
-    // create the channels through which our download results will be passed
-    let (tx, rx) = mpsc::channel();
+    // Limit to an arbitrarily chosen maximum number of parallel downloads.
+    let sem = Arc::new(Semaphore::new(10));
 
     for target in targets.keys() {
-        let repo = repo.clone();
-        let tx = tx.clone();
         info!("Downloading target: {}", target.raw());
-        let target = target.clone();
-        thread_pool.spawn(move || {
-            tx.send(download_targets(&repo, target))
-                // inability to send on this channel is unrecoverable
-                .unwrap();
-        });
+        let cloned_sem = sem.clone();
+        let cloned_repo = repo.clone();
+        let cloned_target = target.clone();
+        let fut = async move {
+            let _permit = cloned_sem.acquire().await.context(error::SemaphoreSnafu)?;
+            download_target(cloned_repo, cloned_target).await
+        };
+        tasks.spawn(fut);
```

Comment:
This could spawn an unbounded number of tasks. Before reading this code, my understanding was that if we are using the multi-threaded runtime this could therefor result in an unbounded number of OS threads being spawned. After re-reading the [runtime docs](https://docs.rs/tokio/latest/tokio/runtime/index.html#multi-thread-scheduler) it seems like the number of OS threads is bound to the number of CPU cores.

In the past I've used [`buffered_unordered`](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.buffer_unordered) to limit the number of tasks as well, which seems slightly safer but it is certainly more complex.